### ISSUE TITLE
Proposal: Parameter extraction / JSON validation API

### ISSUE THREAD

**Issue by sberan** [ingress/router#4]
## Parameter Extraction

Parameter extraction will allow a route handler to fetch specific parameter values from the `Context` object. This is required for validations. We will have decorators for each parameter type (`Path`, `Query`, `Body`), namespaced in the `Param` namespace, as follows:

```typescript
@Route('users')
class MyController {
  @Route.Get(':id')
  getUser(@Param.Path('id') id: number) {
    ...handler body
  }
}
```

In this example `@Param.Path('id')` instructs ingress to pass the `id` path parameter as the first parameter to the handler.

### Simple Type Validation

For primitive types, `@Param` decorators can be used to extract type information about the desired type of the parameter. Type information will be validated and coerced using the json schema library [ajv](https://github.com/epoberezkin/ajv). If the parameter cannot be coerced to the specified type, the user defined error handler is invoked.

### Complex type validation

For complex parameter types, the simple option would be to use classes defined by the user, like so: 

```typescript
import { Schema }  from 'ingress/router'

@Schema()
class User {
 name: string
}

@Route('users')
class MyController {
  @Route.Post()
  createUser(@Param.Body() user: User) {
    console.log(user.name) //succeeds
    console.log(user.asdf) // compile error
  }
}
```
The main advantage of classes for validation is simplicity. They are simple for most developers to understand, as similar validation frameworks exist in other languages. Also, they have a runtime and a compile-time type, and thus can be used to determine the type of a parameter.

The problem with this approach is that classes cannot represent all valid JSON values. For example, top-level arrays cannot be represented using classes. Many property validation types  such as unions would also be impossible to inspect at runtime using this schema style.

Instead, I propose using a thin wrapper around [io-ts](https://github.com/gcanti/io-ts) to define schemas for complex parameter validation.

In an ideal world, we would be able to define a schema like so:

```typescript
import { Validation, schema }  from 'ingress/router'

const UserSchema = schema.object({
  name: schema.string
  dob: schema.date
})

@Route('users')
class MyController {
  @Route.Post()
  createUser(@Param.Body() user: Validation<typeof UserSchema>) {
    console.log(user.name) //succeeds
    console.log(user.asdf) // compile error
  }
}
```

This syntax is pretty good, and can represent any type in typescript. Unfortunately, we cannot introspect the type of `Validation<UserSchema>` at runtime, since the generic type parameter will be lost by the typescript compiler. Instead we'll need to create a validator class which holds the type information, like so:

Final Syntax:

```typescript
import { Validator, schema }  from 'ingress/router'

const User = schema.object({
  name: schema.string,
  dob: schema.date
})

class UserValidator extends Validator<typeof User> {
  getType() { return User }
}

@Route('users')
class MyController {
  @Route.Post()
  createUser(@Param.Body() { value: user }: UserValidator) {
    console.log(user.name) //succeeds
    console.log(user.asdf) // compile error
  }
}
```

**Comment by calebboyd** [comment]
I'm really interested in the class "schema" as I've explored it a bit before. Can you elaborate more on this limitation?

>The problem with this approach is that classes cannot represent all valid JSON values. For example, top-level arrays cannot be represented using classes

Something like? 
```typescript
class MyType extends Array {...}
```
We could also extend the api of the parameter and property decorators, such that more of ajv could be taken advantage of

**Comment by sberan** [comment]
```typescript
class MyType extends Array {...}
```

This ignores parameterized types. I couldn't come up with a type-safe way to represent things like`Array<string>`, `Array<User>`, `Array<string | undefined>` etc, to say nothing of tuples, or array properties on objects. io-ts does all this and more. Keep in mind that these things need to be represented as typescript types that can be introspected at runtime.

**Comment by sberan** [comment]
UPDATE: I was able to get the following syntax to work by figuring out how the typescript compiler deduces parameter types. I'm pretty happy with it! Thoughts?


```
const User = object({
  name: string,
  things: array(string),
  dob: string,
  addresses: array(
    object({
      line1: string,
      line2: union([string, undefined]),
      city: string,
      state: string,
      zip: string
    })
  )
})

type User = Valid<typeof User>

class MyClass {
  @Path('person')
  someMethod (@Body() user: User) {
    user.name
    user.naem                             //error: property "naem" does not exist
    user.things[0].toUpperCase()
    user.dob.substring(0, 2)
    user.addresses[0].line1.toUpperCase()
    user.addresses[0].line2.toUpperCase() //error: [ts] Object is possibly 'undefined'
  }
}
```

**Comment by calebboyd** [comment]
Closed via #6
